storage-proofs-porep/src/stacked/vanilla/proof.rs

```diff
@@ -493,43 +491,107 @@ impl<'a, Tree: 'static + MerkleTreeTrait, G: 'static + Hasher> StackedDrg<'a, Tr
+                                    let layer_data: &mut Vec<_> = &mut layer_data;
+
+                                    // gather all layer data in parallel.
+                                    s.spawn(move |_| {
+                                        for (layer_index, layer_elements) in
+                                            layer_data.iter_mut().enumerate()
+                                        {
+                                            let store = labels.labels_for_layer(layer_index + 1);
+                                            let start = (i * nodes_count) + node_index;
+                                            let end = start + chunked_nodes_count;
+                                            let elements: Vec<<Tree::Hasher as Hasher>::Domain> =
+                                                store
+                                                    .read_range(Range { start, end })
+                                                    .expect("failed to read store range");
+                                            layer_elements
+                                                .extend(elements.into_iter().map(Into::into));
+                                        }
+                                    });
+                                });
+                                // Copy out all layer data arranged into columns.
+                                for layer_index in 0..layers {
+                                    for index in 0..chunked_nodes_count {
+                                        columns[index][layer_index] =
+                                            layer_data[layer_index][index];
+                                    }
                                 }
-                            }
 
-                            drop(layer_data);
+                                columns
+                            };
+
+                            #[cfg(feature = "gpu2")]
+                            let columns: Vec<
+                                GenericArray<Fr, ColumnArity>,
+                            > = {
+                                use fr32::bytes_into_fr;
+
+                                // Allocate layer data array and insert a placeholder for each layer.
+                                let mut layer_data: Vec<Vec<u8>> =
+                                    vec![
+                                        vec![0u8; chunked_nodes_count * std::mem::size_of::<Fr>()];
+                                        layers
+                                    ];
+
+                                rayon::scope(|s| {
+                                    // capture a shadowed version of layer_data.
+                                    let layer_data: &mut Vec<_> = &mut layer_data;
+
+                                    // gather all layer data in parallel.
+                                    s.spawn(move |_| {
+                                        for (layer_index, mut layer_bytes) in
```

Comment:
I think you may want to spawn in the iteration.The current implemention seems not paralleled actually.